lib/Sema/CSDiagnostics.h

```diff
@@ -667,6 +667,22 @@ class AssignmentFailure final : public FailureDiagnostic {
   Optional<OverloadChoice> getMemberRef(ConstraintLocator *locator) const;
 };
 
+class NonClassTypeToAnyObjectConversionFailure : public FailureDiagnostic {
```

Comment:
Let's make this `final` and inherit from `ContextualFailure` this was you can just pass lhs, rhs to the constructor and it would resolve them properly (the reason why it's important is that fix could have type variables in the types which needs to be resolved based on the deduced solution), and you can use `ContextualMismatch::diagnoseAs{Error, Note}();` it your diagnose methods.